src/passes/InstrumentBranchHints.cpp

```diff
@@ -0,0 +1,334 @@
+      //
+      //  (local.set $temp ..)
+      //  (call LOG_BRANCH (local.get $temp))  ;; used in logging
+      //  (if
+      //    (local.get $temp)                  ;; and used in condition
+      //
+      // We also consider the fallthrough, for the nested case:
+      //
+      //  (if
+      //    (block
+      //      (local.set $temp ..)
+      //      (call LOG_BRANCH (local.get $temp))  ;; used in logging
+      //      (local.get $temp)                    ;; and used in condition
+      //    )
+      //
+      // We also consider a tee:
+      //
+      //  (call LOG_BRANCH (local.tee $temp (..)))  ;; used in logging
+      //  (if
+      //    (local.get $temp)                       ;; and used in condition
+      //
+      auto* fallthrough = Properties::getFallthrough(
+        curr->condition, getPassOptions(), *getModule());
+      auto* get = fallthrough->template dynCast<LocalGet>();
+      if (!get) {
+        return;
+      }
+      auto& sets = localGraph->getSets(get);
+      if (sets.size() != 1) {
+        return;
+      }
+      auto* set = *sets.begin();
+      auto& gets = localGraph->getSetInfluences(set);
+      Call* call = nullptr;
+      if (gets.size() == 2) {
+        // The set has two gets: the get in the condition we began at, and
+        // another.
+        LocalGet* otherGet = nullptr;
+        for (auto* get2 : gets) {
+          if (get2 != get) {
+            otherGet = get2;
+          }
+        }
+        assert(otherGet);
+        // See if that other get is used in a logging.
+        auto iter = getsOfPriorInstrumentation.find(otherGet);
+        if (iter == getsOfPriorInstrumentation.end()) {
+          return;
+        }
+        // Great, this is indeed a prior instrumentation.
+        call = iter->second;
+      } else if (gets.size() == 1) {
+        // The set has only one get, but it might be a tee that flows into a
+        // call.
+        auto iter = teesOfPriorInstrumentation.find(set);
+        if (iter == teesOfPriorInstrumentation.end()) {
+          return;
+        }
+        // Great, this is indeed a prior instrumentation.
+        call = iter->second;
+      } else {
+        // The get has more uses; give up, as the pattern is not what we
+        // expect.
+        return;
+      }
+
+      // We found a potential call from a prior instrumentation. It should have
+      // a const ID.
+      assert(call->operands.size() == 3);
+      auto* c = call->operands[0]->template dynCast<Const>();
+      if (!c) {
+        return;
+      }
+      // Emit logging to pair with it, with negated ID.
+      id = -c->value.geti32();
+      if (id > 0) {
+        // The seen ID was already negated, so we negated it again to be
+        // positive. That means the existing instrumentation was a second
+        // instrumentation, and we should only operate on positive IDs and emit
+        // negative ones.
+        return;
+      }
```

Comment:
It seems simpler to check whether the original value is negative before negating it.